vault/identity_store_oidc_provider.go

```diff
@@ -1060,3 +1179,461 @@ func (i *IdentityStore) pathOIDCProviderExistenceCheck(ctx context.Context, req
 	return entry != nil, nil
 }
+
+func (i *IdentityStore) pathOIDCAuthorize(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	ns, err := namespace.FromContext(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("%s: internal server error", ErrAuthServerError)
+	}
+
+	// Get the OIDC provider
+	name := d.Get("name").(string)
+	provider, err := i.getOIDCProvider(ctx, req.Storage, name)
+	if err != nil {
+		return nil, fmt.Errorf("%s: internal server error", ErrAuthServerError)
+	}
+	if provider == nil {
+		return logical.ErrorResponse("%s: provider not found", ErrAuthInvalidRequest), nil
+	}
+
+	// Validate that a scope parameter is present and contains the openid scope value
+	scopes := strutil.ParseStringSlice(d.Get("scope").(string), " ")
+	if len(scopes) == 0 || !strutil.StrListContains(scopes, openIDScope) {
+		return logical.ErrorResponse("%s: scope parameter must contain the %q value",
+			ErrAuthInvalidRequest, openIDScope), nil
+	}
+
+	// Validate the response type
+	responseType := d.Get("response_type").(string)
+	if responseType == "" {
+		return logical.ErrorResponse("%s: response_type parameter is required",
+			ErrAuthInvalidRequest), nil
+	}
+	if responseType != "code" {
+		return logical.ErrorResponse("%s: unsupported response_type value",
+			ErrAuthUnsupportedResponseType), nil
+	}
+
+	// Validate the client ID
+	clientID := d.Get("client_id").(string)
+	if clientID == "" {
+		return logical.ErrorResponse("%s: client_id parameter is required",
+			ErrAuthInvalidClientID), nil
+	}
+	client, err := i.clientByID(ctx, req.Storage, clientID)
+	if err != nil {
+		return nil, fmt.Errorf("%s: internal server error", ErrAuthServerError)
+	}
+	if client == nil {
+		return logical.ErrorResponse("%s: client with client_id not found",
+			ErrAuthInvalidClientID), nil
+	}
+	if !strutil.StrListContains(provider.AllowedClientIDs, "*") &&
+		!strutil.StrListContains(provider.AllowedClientIDs, clientID) {
+		return logical.ErrorResponse("%s: client is not authorized to use the provider",
+			ErrAuthUnauthorizedClient), nil
+	}
+
+	// Validate the redirect URI
+	redirectURI := d.Get("redirect_uri").(string)
+	if redirectURI == "" {
+		return logical.ErrorResponse("%s: redirect_uri parameter is required",
+			ErrAuthInvalidRequest), nil
+	}
+	if !strutil.StrListContains(client.RedirectURIs, redirectURI) {
+		return logical.ErrorResponse("%s: redirect_uri is not allowed for the client",
+			ErrAuthInvalidRedirectURI), nil
+	}
+
+	// Validate the state
+	state := d.Get("state").(string)
+	if state == "" {
+		return logical.ErrorResponse("%s: state parameter is required", ErrAuthInvalidRequest), nil
+	}
+
+	// Validate the nonce
+	nonce := d.Get("nonce").(string)
+	if nonce == "" {
+		return logical.ErrorResponse("%s: nonce parameter is required", ErrAuthInvalidRequest), nil
+	}
+
+	// Validate that there is an identity entity associated with the request
+	entity, err := i.MemDBEntityByID(req.EntityID, false)
```

Comment:
Is it possible for the Entity to exist in Storage but not MemDB? If yes, would that be an issue here?

Reply:
Good question! Dug into this a bit to find out that it's not likely for that to happen (if it did, then it'd suggest a bug in the entity code). The reason is that updates to entities write to memdb _and_ storage (see: [handleEntityUpdateCommon](https://github.com/hashicorp/vault/blob/main/vault/identity_store_entities.go#L272), [upsertEntity](https://github.com/hashicorp/vault/blob/main/vault/identity_store_util.go#L482-L503)). The memdb transaction will be aborted if the storage write fails, which effectively rolls back a partial failure. This is a bit of a different strategy than our client update handler, where we simply invalidate the memdb entry for writes. I think either approach would work.

I think we're safe to use `MemDBEntityByID` here. You can see that the entity read operation uses the same method ([pathEntityIDRead](https://github.com/hashicorp/vault/blob/main/vault/identity_store_entities.go#L322-L340)).

Reply:
Thinking on this more, if we followed a similar pattern (client updates write to both memdb and storage), then we could probably list all clients from memdb for the `allowed_client_ids="*"` case instead of listing from storage. I think that'd be a nice optimization. Will add a stretch task for this and proceed with what we have here.

Reply:
Nice! Thanks for the clarification